### PR TITLE
Supports multiple DCL Numbers for each MARC record entity

### LINKED ISSUE #233: Don't assume bibids are unique
One MARC record in Voyager can be linked to multiple DCLNums, so we shouldn't use them as the local system identifier when importing from Plum.

There may be other places where we assume a 1-to-1 mapping between MARC records and DCLNums.

### PR DESCRIPTION
**Pull request by jrgriffiniii**
Resolves #233 by addressing the following:
- Implements `MarcRecord` as an `ActiveRecord` Model and Assigns a one-to-one relationship between `Books` and `MarcRecords`
- Delegates Digital Cicognara Library (DCL) Numbers for `Books` to `MarcRecords`
Ensures that there can be multiple DCL Numbers per `MarcRecord` (and, by relation, `Book`)
- Ensures that ARKs are indexed for `MarcRecords` and `Books`
- Uses XPaths to MARC XML elements to resolve individual MARC record entities
- Adds Rake tasks for purging the Solr collection of Documents and running a Solr development
environment

### CODE REVIEW

**Comment by escowles** [comment]
I haven't been able to get this working, either on my local machine or in staging.  Locally, when I try to index the full TEI/MARC data files, it takes a long time and eventually dies with the message `Killed: 9`.  So I'm not sure what's going on.

Are you able to index the data files from https://github.com/pulibrary/cicognara-catalogo ?

**Comment by jrgriffiniii** [comment]
I've rebased this and found that this seems to be where memory problems arise: https://github.com/pulibrary/cicognara-rails/blob/master/lib/cicognara/tei_indexer.rb#L99

This in itself does not kill the process, but it drastically increases the memory footprint once all of the XML Elements have been loaded into memory.  My suspicion is that my introduction of  `MarcRecord.resolve` here: https://github.com/pulibrary/cicognara-rails/blob/issues-233-jrgriffiniii-books-multiple-bibids-rebase/lib/cicognara/tei_indexer.rb#L125 is likely where the process crashes

**Comment by jrgriffiniii** [comment]
I'm not comfortable with aspects of this, but locally https://github.com/pulibrary/cicognara-rails/pull/320/commits/af92172fcae0680f47c9f94924c0068e4117e05d resolves the memory leak issue which prevented all records in the Catalogo from being indexed.

File: lib/cicognara/catalogo_item.rb [hunk 1]
```diff
@@ -22,52 +22,57 @@ def solr_doc
...
       end
-      # fields to ignore
-      remove_display_fields(book_fields)
 
-      # single-valued fields
-      field_first_value(book_fields)
+      def marc_fields
+        book_fields = {}
+        books.each do |book|
+          book_doc = book.to_solr
```
Comment (jrgriffiniii) [inline lib/cicognara/catalogo_item.rb]: This is where Solr is now failing to index the MarcRecord object linked to the Book object

File: lib/cicognara/catalogo_item.rb [hunk 2]
```diff
@@ -22,52 +22,57 @@ def solr_doc
...
+          next if book_doc.nil?
+          book_fields.merge!(book_doc) do |_, oldval, newval|
+            values = Array.wrap(newval) + Array.wrap(oldval)
```
Comment (jrgriffiniii) [inline lib/cicognara/catalogo_item.rb]: This also is not properly merging MARC fields extracted from more than one book (in cases where Entries link to more than one Book)

File: lib/cicognara/bulk_entry_indexer.rb [hunk 3]
```diff
@@ -1,16 +1,113 @@
...
-    def initialize(entries)
+    def initialize(entries, index)
       @entries = entries
+      @index = index
+      @entries_cache = {}
+      @books_cache = {}
+    end
+
+    def entry_indexers
+      @entry_indexers ||= @entries.map { |entry| EntryIndexer.new(entry, @index, @entries_cache, @books_cache) }
```
Comment (jrgriffiniii) [inline lib/cicognara/bulk_entry_indexer.rb]: MarcRecord objects (retrieved from Book objects linked to Entry objects) have the full XML documents accessible using `#source` here

File: lib/cicognara/bulk_entry_indexer.rb [hunk 4]
```diff
@@ -1,16 +1,113 @@
...
+    attr_reader :entry
+    def initialize(entry, index, cache, books_cache)
+      @entry = entry
+      @index = index
+      @cache = cache
+      @books_cache = books_cache
+    end
+
+    def books
+      @books ||= @entry.books.uniq(&:id)
```
Comment (jrgriffiniii) [inline lib/cicognara/bulk_entry_indexer.rb]: MarcRecord objects (retrieved from Book objects linked to Entry objects) do *not* have the full XML documents accessible using `#source` here (maybe this is due to `#uniq`?)

File: lib/cicognara/book_indexer.rb [hunk 5]
```diff
@@ -6,45 +6,64 @@ def initialize(books)
...
+
+      def marc_documents
+        marc_docs = books.map(&:marcxml)
+        marc_docs.map(&:clone)
+      end
+
+      def marc_collection(document)
+        collection = Nokogiri::XML('<collection></collection>')
+        return unless document.root.present?
+        collection.root << document.root
```
Comment (jrgriffiniii) [inline lib/cicognara/book_indexer.rb]: This is probably related to the problem.

**Comment by tpendragon** [comment]
It's been a long time. Do we still need this feature?

**Comment by escowles** [comment]
I don't think this is a priority for this sprint — the project has decided to revise a set of MARC records (the first 1000) and work on resolving data quality problems with that subset. As part of that, we can determine if this issue needs to be addressed, or if the current data model is sufficient for the functionality we need.

**Comment by tpendragon** [comment]
Great. I'm gonna close this then.

**Pull request closed without merging**